runtime/bajun/src/lib.rs

```diff
@@ -379,13 +375,75 @@ impl frame_system::Config for Runtime {
 	/// The Block provider type
 	type Block = Block;
 	type RuntimeTask = RuntimeTask;
-	type SingleBlockMigrations = ();
-	type MultiBlockMigrator = ();
+	type SingleBlockMigrations = SingleBlockMigrations;
+	type MultiBlockMigrator = pallet_migrations::Pallet<Runtime>;
 	type PreInherents = ();
 	type PostInherents = ();
 	type PostTransactions = ();
 }
 
+type SingleBlockMigrations = (pallet_ajuna_awesome_avatars::migration::v6::MigrateToV6<Runtime>,);
+
+#[cfg(not(feature = "runtime-benchmarks"))]
+use mbm::MultiBlockMigrations;
+
+#[cfg(not(feature = "runtime-benchmarks"))]
+mod mbm {
+	use crate::Runtime;
+	use pallet_ajuna_awesome_avatars::migration::v6::mbm::{
+		LazyMigrationAvatarV5ToV6, LazyMigrationPlayerSeasonConfigsV5ToV6,
+		LazyMigrationSeasonStatsV5ToV6, LazyTradeStatsMapCleanup,
+	};
+
+	use crate::weights::pallet_ajuna_awesome_avatars_mbm::WeightInfo as AaaMbmWeight;
+
+	pub type MultiBlockMigrations = (
+		LazyMigrationPlayerSeasonConfigsV5ToV6<Runtime, AaaMbmWeight<Runtime>>,
+		LazyMigrationSeasonStatsV5ToV6<Runtime, AaaMbmWeight<Runtime>>,
+		LazyMigrationAvatarV5ToV6<Runtime, AaaMbmWeight<Runtime>>,
+		LazyTradeStatsMapCleanup<Runtime, AaaMbmWeight<Runtime>>,
+	);
+}
+
+parameter_types! {
+	pub MbmServiceWeight: Weight = Perbill::from_percent(80) * RuntimeBlockWeights::get().max_block;
+}
+
+impl pallet_migrations::Config for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	type CursorMaxLen = ConstU32<65_536>;
+	type IdentifierMaxLen = ConstU32<256>;
+	type MigrationStatusHandler = LoggerMigrationStatusHandler;
+	type FailedMigrationHandler = frame_support::migrations::FreezeChainOnFailedMigration;
+	type MaxServiceWeight = MbmServiceWeight;
+	type WeightInfo = weights::pallet_migrations::WeightInfo<Runtime>;
+	#[cfg(feature = "runtime-benchmarks")]
+	type Migrations = pallet_migrations::mock_helpers::MockedMigrations;
+	#[cfg(not(feature = "runtime-benchmarks"))]
+	type Migrations = MultiBlockMigrations;
+}
+
+/// Records all started and completed upgrades in `UpgradesStarted` and `UpgradesCompleted`.
+pub struct LoggerMigrationStatusHandler;
+impl MigrationStatusHandler for LoggerMigrationStatusHandler {
+	fn started() {
+		log::info!("MigrationStatusHandler started");
+	}
+
+	fn completed() {
+		log::info!("MigrationStatusHandler completed");
+	}
+}
+
+/// Records all failed upgrades in `UpgradesFailed`.
+pub struct MockedFailedMigrationHandler;
+impl FailedMigrationHandler for MockedFailedMigrationHandler {
+	fn failed(migration: Option<u32>) -> FailedMigrationHandling {
+		log::error!("FailedMigrationHandler failed at: {migration:?}");
+		FailedMigrationHandling::ForceUnstuck
```

Comment:
This will make the chain resume normal operation and allow extrinsics again, but it will not continue the migration. Here, I suggest we assume that it works now after testing with the production data. Otherwise we could start integrating the [pallet-safe-mode](https://github.com/paritytech/polkadot-sdk/blob/305d311d5c732fcc4629f3295768f1ed44ef434c/substrate/frame/safe-mode/src/lib.rs), which is not audited yet though.

Reply:
I suggest keeping it at `ForceUnstuck`. The only thing that could go wrong here is that a few entries are messed up, but no system logic is touched. `pallet-safe-mode` can be evaluated at another time.

Reply:
To me it looks like you are actually using the freeze option:  
`type FailedMigrationHandler = frame_support::migrations::FreezeChainOnFailedMigration;`.

Reply:
ooops, good catch. Fuck up on my end. Thanks for pointing this out!